### PR TITLE
[MRG+1] Add bep006 eeg

### LINKED ISSUE #34: BIDS-EEG in MNE-BIDS
Probably better if this has its own issue.

### PR DESCRIPTION
**Pull request by sappelhoff**
This will introduce BEP_006 "EEG" to MNE-BIDS and close #34 

Current state of the BEP_006: [Google doc](https://docs.google.com/document/d/1ArMZ9Y_quTKXC-jNXZksnedK2VHHoKP3HCeO5HPcgLE)

BEP_006 Examples: https://github.com/incf/bids-examples/tree/bep006_eeg

The bids-validator can already deal with EEG data by including the bep006 flag such as: `bids-validator --bep006 /path/to/dataset`

# To Do:
- [x] Make MNE-BIDS aware that there are more types of data next to MEG (contingent on #73)
    - for now, keep using `kind`
- [x] Introduce all EEG supported file types and test that they are converted and written correctly
    - BrainVision is a multi-file format: .vhdr, .eeg, .vmrk
    - EEGLAB .set files might be accompanied by a .fdt file
- [x] Make an example for EEG to be posted in the [gallery](https://mne-tools.github.io/mne-bids/auto_examples/index.html)
- [x] skip `coordsystem.json` and `electrodes.tsv` for EEG until the handling of coordinates is resolved in MNE-Python (i.e., no rescaling of the coordinates read from a file: https://github.com/mne-tools/mne-python/pull/5472)
- [x] `scans.tsv` is equal across modalities, write it for all (so far it was only written for meg)
- [x] `events.tsv` has been adjusted already in #55, no changes
- [x] reformat `sidecar.json` to better cover all shared keys. Issue for EEG REQUIRED: `EEGReference` ... make it `n/a` until we find a good way to infer it from mne.raw
- [x] find solution for EDF data
    - for `mne.find_events`, it wants a parameter `initialEvent=True`
    - when being read, it wants `preload=True`

------------
# General list of things to implement
### Must-haves in the future:
- `coordsystem.json` for EEG
- `electrodes.tsv` for EEG

### Nice to haves ... but not too important and hard to implement
... for these, I currently do not have an idea how to extract the information automatically. Adding them as optional (None-defaulting) parameters would clutter the API - so I don't work on them for now.
- Several `sidecar.json` fields in decreasing order of importance
    - HardwareFilters
    - ManufacturersModelName
    - CapManufacturer
    - CapManufacturersModelName

--> as a potential solution, see #47

### CODE REVIEW

**Comment by jasmainak** [comment]
@sappelhoff I didn't look closely but PR looks clean so far. Could you keep the diff small so it's easy to review? Thanks!

**Comment by sappelhoff** [comment]
So the skeleton to include EEG is ready. Please make a first round of reviews @jasmainak @teonbrooks @monkeyman192 so that I know what you consider a total no-go or for what you have better ideas.

What is still missing is
1. The example (ignore it for now - this is outdated)
2. Tests of
    - the three new functions in `utils.py` I introduce
    - actually converting one of each .vhdr, .edf, .set, .bdf, .cnt

PS:
> Could you keep the diff small so it's easy to review? 

This is really hard, because MNE-BIDS is quite MEG-centric and several changes have to be made to generalize to the other modalities

**Comment by jasmainak** [comment]
@sappelhoff what I'd really like to see even before examples is tests ...

File: examples/make_eeg_bids.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+#
+#    ---------- CHANGES
+#    ---------- dataset_description.json
+#    ---------- participants.tsv
+#    ---------- README
+#    ---------- sub-xx
+#        ---------- modality
+#            ---------- sub-xx_task-yy_modality.ext
+#            ---------- sub-xx_task-yy_modality.json
+#            ---------- sub-xx_task-yy_events.tsv
```
Comment (jasmainak) [inline examples/make_eeg_bids.py]: I would suggest removing L139--150. L153 will basically take care of it

File: examples/make_eeg_bids.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+###############################################################################
+# We are importing everything we need for this example:
+import os
+
+from mne import find_events
+from mne.io import read_raw_brainvision
+
+from mne_bids import raw_to_bids
+from mne_bids.datasets import download_matchingpennies_subj
```
Comment (jasmainak) [inline examples/make_eeg_bids.py]: this is not yet there.

Also, if I remember correctly, this dataset was already in BIDS format. How do you propose to use it in the example in that case?

**Comment by jasmainak** [comment]
Diff looks clean so far :) Tests next.

File: examples/make_eeg_bids.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+# * '.vhdr' for header information
+# * '.vmrk' for specifying the event markers within the data.
+
+###############################################################################
+# Step 2: Formatting as BIDS
+# --------------------------
+#
+# We are reading the data using MNE-Python's io module and the
+# `read_raw_brainvision` function, which takes a `.vhdr` file as an input.
+vhdr_path = os.path.join(data_dir, 'sub-05_task-matchingpennies_eeg.vhdr')
```
Comment (jasmainak) [inline examples/make_eeg_bids.py]: actually thinking a bit more, why not add an example for the `physionet` data: https://martinos.org/mne/dev/generated/mne.datasets.eegbci.load_data.html ? The fetcher already exists and there are many subjects (we could just fetch 2 or 3) -- so we can demonstrate how to do the conversion for more than n=1 subjects.

Comment (sappelhoff) [reply examples/make_eeg_bids.py]: This is an excellent idea, I'm on it!

**Comment by monkeyman192** [comment]
wrt the extra stuff in the `sidecar.json`, see #47 
I think this is still something that could be good to do and would allow pretty much any extra data we want to add.

**Comment by sappelhoff** [comment]
EEGLAB bug needs to be solved to continue: mne-tools/mne-python#5503

**Comment by jasmainak** [comment]
okay I see! Feel free to ping me if you need a review on the PR to fix the bug

File: mne_bids/mne_bids.py [hunk 4]
```diff
@@ -253,7 +271,7 @@ def _scans_tsv(raw, raw_fname, fname, verbose):
     """
     # get MEASurement date from the data info
     meas_date = raw.info['meas_date']
-    if isinstance(meas_date, (np.ndarray, list)):
+    if isinstance(meas_date, (np.ndarray, list, tuple)):
         meas_date = meas_date[0]
```
Comment (sappelhoff) [inline mne_bids/mne_bids.py]: In master of MNE-Python, `meas_date` has just been formatted to consistently be a tuple. So soon, we will be able to get rid of np.ndarray and list over there. Should I make a comment so we don't forget? https://github.com/mne-tools/mne-python/pull/5500

File: mne_bids/tests/test_mne_bids.py [hunk 5]
```diff
@@ -130,3 +130,75 @@ def test_bti():
...
+    raw_to_bids(subject_id=subject_id, session_id=session_id, run=run,
+                task=task, raw_file=raw_fname, output_path=output_path,
+                overwrite=True, kind='eeg')
+
+    cmd = ['bids-validator', '--bep006', output_path]
+    run_subprocess(cmd, shell=shell)
+
+
+def test_set():
+    """Test raw_to_bids conversion for EEGLAB data."""
```
Comment (sappelhoff) [inline mne_bids/tests/test_mne_bids.py]: This will only pass on master of MNE-Python, once https://github.com/mne-tools/mne-python/pull/5504#issuecomment-420041448 is merged

File: examples/make_eeg_bids.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
+"""
+=========================
+Use MNE-Bids for EEG Data
```
Comment (teonbrooks) [inline examples/make_eeg_bids.py]: nit: Bids -> BIDS

File: examples/make_eeg_bids.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+=========================
+
+In this example, we use MNE-Bids to create a BIDS-compatible directory of EEG
```
Comment (teonbrooks) [inline examples/make_eeg_bids.py]: here too

File: examples/make_eeg_bids.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+data_dir = os.path.join(data_dir, 'physiobank', 'database', 'eegmmidb')
+edf_path = os.path.join(data_dir, 'S001', 'S001R02.edf')
+raw = read_raw_edf(edf_path, preload=True)
+print(raw)
+
+###############################################################################
+# With this simple step we have everything to start a new BIDS directory using
+# our data. To do that, we can use the high level function `raw_to_bids`, which
+# is the core of MNE-BIDS. Generally, `raw_to_bids` tries to extract as much
+# meta data as possible from the raw data and then format it in a BIDS
```
Comment (teonbrooks) [inline examples/make_eeg_bids.py]: format -> formats

File: mne_bids/mne_bids.py (hunk 4 above)
Comment (teonbrooks) [reply mne_bids/mne_bids.py]: I would make a comment and add an issue for the mne version that we will freeze for that particular release

File: mne_bids/mne_bids.py [hunk 9]
```diff
@@ -586,12 +632,19 @@ def raw_to_bids(subject_id, task, raw_file, output_path, session_id=None,
         raise ValueError('ext must be in %s, got %s'
                          % (''.join(ALLOWED_EXTENSIONS), ext))
 
-    # for FIF, we need to re-save the file to fix the file pointer
-    # for files with multiple parts
+    # Copy the imaging data files
+    # Re-save FIF files to fix the file pointer for files with multiple parts
```
Comment (teonbrooks) [inline mne_bids/mne_bids.py]: we should a comment that this is not currently implemented. I am working on it with this pr: https://github.com/mne-tools/mne-python/pull/5470

File: mne_bids/tests/test_mne_bids.py [hunk 10]
```diff
@@ -130,3 +130,75 @@ def test_bti():
...
+    output_path = '/home/stefanappelhoff/Desktop/bogus'  # _TempDir()
```
Comment (teonbrooks) [inline mne_bids/tests/test_mne_bids.py]: guessing this path is temp for now?

File: mne_bids/tests/test_utils.py [hunk 11]
```diff
@@ -71,3 +76,89 @@ def test_age_on_date():
     assert age_on_date(bday, exp3) == 24
     with pytest.raises(ValueError):
         age_on_date(bday, exp4)
+
+
+def test_get_brainvision_paths():
```
Comment (teonbrooks) [inline mne_bids/tests/test_utils.py]: you didn't double underscore ;)

File: mne_bids/tests/test_utils.py [hunk 12]
```diff
@@ -71,3 +76,89 @@ def test_age_on_date():
...
+def test_copyfile_brainvision():
+    """Test the copying of BrainVision vhdr, vmrk and eeg files."""
+    output_path = _TempDir()
+    data_path = op.join(base_path, 'brainvision', 'tests', 'data')
+    raw_fname = op.join(data_path, 'test.vhdr')
+    new_name = op.join(output_path, 'tested_conversion.vhdr')
+
+    # IO error testing
+    with pytest.raises(ValueError):
+        copyfile_brainvision(raw_fname, new_name+'.eeg')
```
Comment (teonbrooks) [inline mne_bids/tests/test_utils.py]: whitespace around operator? not sure if this is a pep8 violation or just style

File: mne_bids/tests/test_utils.py [hunk 13]
```diff
@@ -71,3 +76,89 @@ def test_age_on_date():
...
+def test_copyfile_eeglab():
+    """Test the copying of EEGlab set and fdt files."""
+    output_path = _TempDir()
+    data_path = op.join(testing.data_path(), 'EEGLAB')
+    raw_fname = op.join(data_path, 'test_raw.set')
+    new_name = op.join(output_path, 'tested_conversion.set')
+
+    # IO error testing
+    with pytest.raises(ValueError):
+        copyfile_eeglab(raw_fname, new_name+'.wrong')
```
Comment (teonbrooks) [inline mne_bids/tests/test_utils.py]: same as above

File: mne_bids/utils.py [hunk 14]
```diff
@@ -347,3 +354,198 @@ def _read_events(events_data, raw):
...
+    if not vmrk_file_match:
+        raise ValueError('Could not find a .vmrk file link in'
+                         ' {}'.format(vhdr_path))
+    else:
+        vmrk_file = vmrk_file_match.groups()[0]
+
+    # Make sure we are dealing with file names as is customary, not paths
+    # Paths are problematic when copying the files to another system. Instead,
+    # always use the file name and keep the file triplet in the same directory
+    assert os.sep not in eeg_file
```
Comment (teonbrooks) [inline mne_bids/utils.py]: can these paths be in different folder hierarchies at recording? if so, maybe we should have a better error handling here

File: mne_bids/utils.py [hunk 15]
```diff
@@ -347,3 +354,198 @@ def _read_events(events_data, raw):
...
+        will be given names as in vhdr_dest with adjusted extensions.
+        Internal file pointers will be fixed.
+
+    """
+    # Get extenstion of the brainvision file
+    fname_src, ext_src = _parse_ext(vhdr_src)
+    fname_dest, ext_dest = _parse_ext(vhdr_dest)
+    if ext_src != ext_dest:
+        raise ValueError('Need to move data with same extension'
+                         ' but got {}, {}'.format(ext_src, ext_dest))
```
Comment (teonbrooks) [inline mne_bids/utils.py]: should include double quotes are {} to highlight str extensions

File: mne_bids/utils.py [hunk 16]
```diff
@@ -347,3 +354,198 @@ def _read_events(events_data, raw):
...
+    """Copy a EEGLAB files to a new location and adjust pointer to '.fdt' file.
+
+    Some EEGLAB .set files come with a .fdt binary file that contains the data.
+    When moving a .set file, we need to check for an associated .fdt file and
+    move it to an appropriate location as well as update an internal pointer
+    within the .set file.
+
+    Notes
+    -----
+    NOT IMPLEMENTED YET. This function will abort upon the encounter of a .fdt
```
Comment (teonbrooks) [inline mne_bids/utils.py]: wip

File: mne_bids/mne_bids.py [hunk 17]
```diff
@@ -24,23 +25,40 @@
...
 UNITS = {'.sqd': 'm', '.con': 'm', '.fif': 'm', '.pdf': 'm', '.ds': 'cm'}
 
-MANUFACTURERS = {'.sqd': 'KIT/Yokogawa', '.con': 'KIT/Yokogawa',
-                 '.fif': 'Elekta', '.pdf': '4D Magnes', '.ds': 'CTF',
-                 '.meg4': 'CTF'}
+meg_manufacturers = {'.sqd': 'KIT/Yokogawa', '.con': 'KIT/Yokogawa',
+                     '.fif': 'Elekta', '.pdf': '4D Magnes', '.ds': 'CTF',
+                     '.meg4': 'CTF'}
+
+eeg_manufacturers = {'.vhdr': 'BrainProducts', '.eeg': 'BrainProducts',
```
[the remark this answers is not in the document]
Comment (sappelhoff) [reply mne_bids/mne_bids.py]: If the user does not have a BEP006 compatible file format (BrainVision, EDF, BDF, CNT, SET) anymore but only the resulting FIF files, then she would first have to convert her FIF data to BrainVision e.g., via the BrainVision writer currently hosted on github.com/palday/philistine. --> and then a conversion to BIDS can be done.

File: mne_bids/tests/test_mne_bids.py (hunk 10 above)
Comment (sappelhoff) [reply mne_bids/tests/test_mne_bids.py]: yes, some internal testing - thanks for noting it

File: mne_bids/tests/test_utils.py (hunk 11 above)
Comment (sappelhoff) [reply mne_bids/tests/test_utils.py]: are we doing that? @jasmainak wasn't too impressed when I tried to do it the last time :-D

File: mne_bids/utils.py (hunk 14 above)
Comment (sappelhoff) [reply mne_bids/utils.py]: For both BrainVision and EEGLAB we can have arbitrary pointers to files. When these files are written by e.g., BrainVision Recorder of saved with EEGLAB the file pointers will always be simply the file name and we assume the file that is being pointed to to be in the same directory.

*However*, user might manually change the pointers and include a path instead of only the file name. That is something I have not yet encountered ... but it is technically possible.

**Comment by sappelhoff** [comment]
thanks for the review @teonbrooks !

> looks good so far. are you going to wait til the eeglab files are supported first?

No, I wouldn't: If users run into a problem with that, they can reconvert their `.set`+`.fdt` combination to be only a standalone `.set`. I would make this a separate issue once this PR is merged because solving this will involve vigorous MAT file handling via python.

File: mne_bids/mne_bids.py [hunk 18]
```diff
@@ -13,6 +13,7 @@
 from collections import defaultdict, OrderedDict
 
 import numpy as np
+from mne import Epochs
```
[the remark this answers is not in the document]
Comment (sappelhoff) [reply mne_bids/mne_bids.py]: good point. This is also something that should be discussed in the main spec, where we have the field `RecordingType` with possible values "continuous" and "epoched" ... this might be obsolete if epochs are considered derivatives

File: mne_bids/mne_bids.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (teonbrooks) [reply mne_bids/mne_bids.py]: oh, yeah, i forgot that FIF is not a supported eeg file format. my bad

File: mne_bids/tests/test_utils.py (hunk 12 above)
Comment (sappelhoff) [reply mne_bids/tests/test_utils.py]: mmmh I'd say it's a violation. Curiously, my linter did not catch it (linter-flake8 in atom)

File: mne_bids/tests/test_utils.py (hunk 11 above)
Comment (sappelhoff) [reply mne_bids/tests/test_utils.py]: I tried to find some conventions but the only one i could find was to name the tests [`test_***`](https://pytest.readthedocs.io/en/reorganize-docs/new-docs/user/naming_conventions.html). So I changed the `test__***` to simply be `test_***` ... because that double underscore does look a bit silly

**Comment by sappelhoff** [comment]
Note: This PR will need #84 to run successfully. 

Also, the `test_bti` test will still fail - due to the Bids-validator issue AND our poor handling of BTi data. As I already suggested in #84, I suggest to solve that in a separate PR.

So now this PR is [MRG]

File: mne_bids/tests/test_utils.py (hunk 11 above)
Comment (jasmainak) [reply mne_bids/tests/test_utils.py]: In mne, we use the single underscore :)

**Comment by sappelhoff** [comment]
ready for merge @jasmainak @teonbrooks

File: examples/make_eeg_bids.py [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+# (`BIDS <http://bids.neuroimaging.io/>`).
+#
+# Conveniently, there is already a data loading function available with
+# MNE-Python:
+
+# Make a directory to save the data to
+home = os.path.expanduser('~')
+mne_dir = os.path.join(home, 'mne_data')
+if not os.path.exists(mne_dir):
+    os.makedirs(mne_dir)
```
Comment (jasmainak) [inline examples/make_eeg_bids.py]: why is this necessary?

File: examples/make_eeg_bids.py [hunk 20]
```diff
@@ -0,0 +1,199 @@
...
+
+# Define which tasks we want to download.
+tasks = [2,  # This is 2 minutes of eyes closed rest
+         4,  # This is run #1 of imagining to close left or right fist
+         12]  # This is run #2 of imagining to close left or right fist
```
Comment (jasmainak) [inline examples/make_eeg_bids.py]: too verbose in my opinion but I can live the comments if you insist :)

File: examples/make_eeg_bids.py [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+
+data_dir = os.path.join(mne_dir, 'MNE-eegbci-data')
+print_dir_tree(data_dir)
+
+###############################################################################
+# The data are in the `European Data Format <https://www.edfplus.info/>`_
+# '.edf', which is good for us because next to the BrainVision format, EDF is
+# one of the recommended file formats for EEG BIDS. However, apart from the
+# data format, we need to build a directory structure and supply meta data
+# files to properly *bidsify* this data.
```
Comment (jasmainak) [inline examples/make_eeg_bids.py]: bidsify in italics maybe? :)

File: examples/make_eeg_bids.py [hunk 22]
```diff
@@ -0,0 +1,199 @@
...
+# We loaded 'S001R02.edf', which corresponds to subject 1 in the second task.
+# The second task was to rest with closed eyes.
+subject_id = '001'  # zero padding to account for >100 subjects in this dataset
+task = 'resteyesclosed'
+raw_file = raw
+output_path = os.path.join(home, 'mne_data', 'eegmmidb_bids')
+
+# Clean up existing directories for the sake of the example
+if os.path.exists(output_path):
+    sh.rmtree(output_path)
```
Comment (jasmainak) [inline examples/make_eeg_bids.py]: why is this needed? Is it related to our `overwrite` discussion?

File: examples/make_eeg_bids.py [hunk 23]
```diff
@@ -0,0 +1,199 @@
...
+
+###############################################################################
+# What does our fresh BIDS directory look like?
+print_dir_tree(output_path)
+
+###############################################################################
+# Looks good so far, let's convert the data for all tasks and subjects.
+
+# Start with a clean directory
+sh.rmtree(output_path)
```
Comment (jasmainak) [inline examples/make_eeg_bids.py]: [again](https://github.com/mne-tools/mne-bids/pull/78/files#r217536493) ?

File: examples/make_eeg_bids.py [hunk 24]
```diff
@@ -0,0 +1,199 @@
...
+# Make sure to pay a lot of attention to the automatic output and make
+# adjustments where necessary.
+#
+# Remember that there is a convenient javascript tool to validate all your BIDS
+# directories called the "BIDS-validator", available as a web version and a
+# command line tool:
+#
+# Web version: https://incf.github.io/bids-validator/
+#
+# Command line tool: https://www.npmjs.com/package/bids-validator
```
Comment (jasmainak) [inline examples/make_eeg_bids.py]: very nice! Thanks @sappelhoff for creating this nice example

File: mne_bids/mne_bids.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (jasmainak) [reply mne_bids/mne_bids.py]: I guess we can leave it in there. It's up to the user to ensure if the data they provide is "raw" or not ...

File: mne_bids/mne_bids.py [hunk 25]
```diff
@@ -386,15 +421,22 @@ def _sidecar_json(raw, task, manufacturer, fname, kind,
...
         ('SamplingFrequency', sfreq),
-        ("DewarPosition", "XXX"),
-        ("DigitizedLandmarks", False),
-        ("DigitizedHeadPoints", False),
-        ("SoftwareFilters", "n/a"),
+        ('SoftwareFilters', 'n/a'),
+        ('RecordingDuration', raw.times[-1]),
+        ('RecordingType', rec_type)]
+    ch_info_json_meg = [
+        ('DewarPosition', 'XXX'),
```
Comment (jasmainak) [inline mne_bids/mne_bids.py]: would you mind changing this to 'n/a'? I know it's not related to your PR, but I just noticed it and it's easy enough for you :)

File: mne_bids/mne_bids.py (hunk 9 above)
Comment (jasmainak) [reply mne_bids/mne_bids.py]: @teonbrooks I guess we would need a `copyfile_fif` function eventually? Or is your MNE PR going to take care of it?

File: mne_bids/tests/test_io.py [hunk 26]
```diff
@@ -7,7 +7,7 @@
 from mne_bids.io import _parse_ext, _read_raw
 
 
-def test__parse_ext():
+def test_parse_ext():
```
Comment (jasmainak) [inline mne_bids/tests/test_io.py]: :)

File: examples/make_eeg_bids.py (hunk 19 above)
Comment (sappelhoff) [reply examples/make_eeg_bids.py]: Because I am using `print_dir_tree` on that directory later ... and if I didn't introduce it here, the path would come "out of nowhere" because `eegbci.load_data()` would have picked the default path. I added it for a better flow of the example

File: examples/make_eeg_bids.py (hunk 22 above)
Comment (sappelhoff) [reply examples/make_eeg_bids.py]: yes exactly

File: examples/make_eeg_bids.py (hunk 23 above)
Comment (sappelhoff) [reply examples/make_eeg_bids.py]: yes, we can remove it once we properly `overwrite`

File: examples/make_eeg_bids.py (hunk 21 above)
Comment (sappelhoff) [reply examples/make_eeg_bids.py]: this is already in italics: 
image

File: mne_bids/utils.py [hunk 27]
```diff
@@ -345,5 +352,200 @@ def _read_events(events_data, raw):
...
+    """
+    # How many of the channels in raw are based on the extended 10/20 system
+    # If more than 80 percent, assume it is 10/20
+    montage1005 = read_montage(kind='standard_1005')
+    counter1005 = 0
+    for ch in raw.ch_names:
+        if ch in montage1005.ch_names:
+            counter1005 += 1
+
+    if counter1005 >= int(0.8*len(raw.ch_names)):
```
[the remark this answers is not in the document]
Comment (sappelhoff) [reply mne_bids/utils.py]: Sometimes there are a few other channels like EOG ... and as long as almost all channels are based on the 10/20 system, we can call it 10/20 - the 80% is arbitrary though ...

File: examples/make_eeg_bids.py (hunk 20 above)
Comment (sappelhoff) [reply examples/make_eeg_bids.py]: I do, I like verbose examples ;-)

File: mne_bids/utils.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (jasmainak) [reply mne_bids/utils.py]: but if you do `pick_types(eeg=True, ...)` on `raw.info`, you could enforce all of them to be based on the 10/20 system no?

File: examples/make_eeg_bids.py (hunk 19 above)
Comment (jasmainak) [reply examples/make_eeg_bids.py]: can you share how the link of the built example? So I can understand better

File: examples/make_eeg_bids.py [hunk 28]
```diff
@@ -0,0 +1,198 @@
...
+
+###############################################################################
+# MNE-BIDS has created a suitable directory structure for us, and among other
+# meta data files, it started an `events.tsv` and `channels.tsv` and made an
+# initial `dataset_description` on top!
+#
+# That's nice and it has saved us a lot of work. However, a few things are not
+# yet covered by MNE-BIDS for EEG data and will require some manual tweaking.
+# Make sure to pay a lot of attention to the automatic output and make
+# adjustments where necessary.
```
Comment (jasmainak) [inline examples/make_eeg_bids.py]: Instead of putting a disclaimer, I would suggest pointing the user to look for `'n/a'` entries in the data to fill them up themselves. Is there actually a possibility of MNE-BIDS putting any information incorrectly? We don't do any guessing, right?

File: examples/make_eeg_bids.py (hunk 19 above)
Comment (jasmainak) [reply examples/make_eeg_bids.py]: okay I see what you mean, but `eegbci.load_data` returns the `path`. Can't we use that?

File: examples/make_eeg_bids.py (hunk 22 above)
Comment (jasmainak) [reply examples/make_eeg_bids.py]: yes but we really need to show an example with incremental addition of data? I would delay it until we've ironed out the `overwrite` API. I don't want to give users ideas ... :)

File: mne_bids/utils.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (jasmainak) [reply mne_bids/utils.py]: I would rather prevent guessing if possible ...

[the remark this answers is not in the document]
Comment (sappelhoff) [reply mne_bids/utils.py]: I agree and will change it

File: examples/make_eeg_bids.py (hunk 28 above)
Comment (sappelhoff) [reply examples/make_eeg_bids.py]: > We don't do any guessing, right?

We don't and we shouldn't - I'll rephrase!

**Comment by sappelhoff** [comment]
There is a new error ... possibly introduced through upstream MNE-Python changes? 

`RuntimeError: By default, MNE does not load data into main memory to conserve resources. adding, dropping, or reordering channels requires raw data to be loaded. Use preload=True (or string) in the constructor or raw.load_data().`

File: examples/make_eeg_bids.py (hunk 19 above)
Comment (jasmainak) [reply examples/make_eeg_bids.py]: so what I'm concerned about is that this path may not always be `mne_data`. See here: https://github.com/mne-tools/mne-python/blob/master/mne/datasets/eegbci/eegbci.py#L27-L32. It depends on your MNE config and I think it's quite common to not store your data in the home directory.

File: examples/make_eeg_bids.py [hunk 29]
```diff
@@ -0,0 +1,197 @@
...
+# Define which tasks we want to download.
+tasks = [2,  # This is 2 minutes of eyes closed rest
+         4,  # This is run #1 of imagining to close left or right fist
+         12]  # This is run #2 of imagining to close left or right fist
+
+# Download the data for subjects 1 and 2
+for subj_idx in [1, 2]:
+    eegbci.load_data(subject=subj_idx,
+                     runs=tasks,
+                     path=mne_dir,
```
Comment (jasmainak) [inline examples/make_eeg_bids.py]: ah you did this. Okay, I can live with it. It's fine ...

File: examples/make_eeg_bids.py (hunk 19 above)
Comment (jasmainak) [reply examples/make_eeg_bids.py]: > As for a link ... now it finally pays off that we started using circleci

yes totally !

**Comment by jasmainak** [comment]
Good to go from my end. @teonbrooks @choldgraf or @agramfort do you want to take a look before I merge?

**Comment by agramfort** [comment]
is it possible to see the rendered doc?

**Comment by sappelhoff** [comment]
yes - It's in the artifacts tab of the circle ci service:

https://119-89170358-gh.circle-artifacts.com/0/html/auto_examples/make_eeg_bids.html

File: examples/make_eeg_bids.py [hunk 30]
```diff
@@ -0,0 +1,197 @@
...
+# Now we have written our BIDS directory.
+print_dir_tree(output_path)
+
+###############################################################################
+# MNE-BIDS has created a suitable directory structure for us, and among other
+# meta data files, it started an `events.tsv` and `channels.tsv` and made an
+# initial `dataset_description` on top!
+#
+# Now it's time to manually check the BIDS directory and the meta files to add
+# all the information that MNE-BIDS could not infer. these places are marked
```
Comment (agramfort) [inline examples/make_eeg_bids.py]: these -> These

File: mne_bids/io.py [hunk 31]
```diff
@@ -51,6 +59,24 @@ def _read_raw(raw_fname, electrode=None, hsp=None, hpi=None, config=None,
...
 
+    # EEG File Types
+    # --------------
+    # BrainVision format by Brain Products, links to  a .eeg and a .vmrk file
+    elif ext == '.vhdr':
+        raw = io.read_raw_brainvision(raw_fname)
+
+    # EDF (european data format) or BDF (biosemi) format
+    elif ext == '.edf' or ext == '.bdf':
+        raw = io.read_raw_edf(raw_fname, preload=True)
```
Comment (agramfort) [inline mne_bids/io.py]: why is using preload=True necessary?

File: examples/make_eeg_bids.py [hunk 32]
```diff
@@ -0,0 +1,197 @@
...
+
+###############################################################################
+# Step 2: Formatting as BIDS
+# --------------------------
+#
+# Let's start by formatting a single subject. We are reading the data using
+# MNE-Python's io module and the `read_raw_edf` function.
+data_dir = os.path.join(data_dir, 'physiobank', 'database', 'eegmmidb')
+edf_path = os.path.join(data_dir, 'S001', 'S001R02.edf')
+raw = read_raw_edf(edf_path, preload=True)
```
Comment (agramfort) [inline examples/make_eeg_bids.py]: same here. Why preload=True?

File: mne_bids/io.py (hunk 31 above)
Comment (sappelhoff) [reply mne_bids/io.py]: Because MNE-Python wants it so for the EDF reader, else you get:

`RuntimeError: EDF+ Annotations (TAL) channel needs to be parsed completely on loading. You must set preload parameter to True.`

File: examples/make_eeg_bids.py (hunk 32 above)
Comment (sappelhoff) [reply examples/make_eeg_bids.py]: see above, this is for all EDF files

File: mne_bids/io.py (hunk 31 above)
Comment (agramfort) [reply mne_bids/io.py]: ok fair enough.

I get we need to fix this in mne-python. We should not need to preload to get the raw.annotations available.

cc @massich

**Comment by jasmainak** [comment]
alright, merged @sappelhoff ! Good way to start your week ;)

File: doc/whats_new.rst [hunk 33]
```diff
@@ -15,6 +15,7 @@ Current
 Changelog
 ~~~~~~~~~
 
+- Add support for EEG and a corresponding example: `make_eeg_bids.py` by `Stefan Appelhoff`_ (`#78 <https://github.com/mne-tools/mne-bids/pull/78>`_)
```
Comment (jasmainak) [inline doc/whats_new.rst]: oops, this should have been added at the bottom. Okay we can fix it later

**Comment by teonbrooks** [comment]
we have EEG support! 🎉 thanks @sappelhoff!
